allennlp/data/iterators/bucket_iterator.py

```diff
@@ -88,16 +88,17 @@ def _sort_dataset_by_padding(dataset: Dataset,
         """
         instances_with_lengths = []
         for instance in dataset.instances:
-            padding_lengths = instance.get_padding_lengths()
+            padding_lengths = cast(Dict[str, Dict[str, float]], instance.get_padding_lengths())
             print("Instance:", instance)
             print("padding lengths:", padding_lengths)
             if padding_noise > 0.0:
                 noisy_lengths = {}
                 for field_name, field_lengths in padding_lengths.items():
                     noisy_lengths[field_name] = add_noise_to_dict_values(field_lengths, padding_noise)
                 padding_lengths = noisy_lengths
-            instance_with_lengths = [padding_lengths[field_name][padding_key]
-                                     for (field_name, padding_key) in sorting_keys] + [instance]
+            instance_with_lengths = ([padding_lengths[field_name][padding_key]
+                                      for (field_name, padding_key) in sorting_keys],
+                                     instance)
             instances_with_lengths.append(instance_with_lengths)
-        instances_with_lengths.sort(key=lambda x: x[:-1])
```

Comment:
that's a side-benefit, I did it because the heterogeneous lists didn't type-check :)